meta/rpc.go

```diff
@@ -0,0 +1,430 @@
+package meta
+
+import (
+	"encoding/binary"
+	"errors"
+	"fmt"
+	"io"
+	"io/ioutil"
+	"log"
+	"net"
+	"time"
+
+	"github.com/gogo/protobuf/proto"
+	"github.com/influxdb/influxdb/meta/internal"
+)
+
+// RPC handles request/response style messaging between cluster nodes
+type RPC struct {
+	Logger         *log.Logger
+	tracingEnabled bool
+
+	store interface {
+		cachedData() *Data
+		IsLeader() bool
+		Leader() string
+		Peers() []string
+		AddPeer(host string) error
+		CreateNode(host string) (*NodeInfo, error)
+		NodeByHost(host string) (*NodeInfo, error)
+		WaitForDataChanged() error
+	}
+}
+
+type JoinResult struct {
+	RaftEnabled bool
+	Peers       []string
+	NodeID      uint64
+}
+
+type Reply interface {
+	GetHeader() *internal.ResponseHeader
+}
+
+// proxyLeader proxies the connection to the current raft leader
+func (r *RPC) proxyLeader(conn *net.TCPConn) {
+	if r.store.Leader() == "" {
+		r.sendError(conn, "no leader")
+		return
+	}
+
+	leaderConn, err := net.DialTimeout("tcp", r.store.Leader(), 10*time.Second)
+	if err != nil {
+		r.sendError(conn, fmt.Sprintf("dial leader: %v", err))
+		return
+	}
+	defer leaderConn.Close()
+
+	leaderConn.Write([]byte{MuxRPCHeader})
+	if err := proxy(leaderConn.(*net.TCPConn), conn); err != nil {
+		r.sendError(conn, fmt.Sprintf("leader proxy error: %v", err))
+	}
+}
+
+// handleRPCConn reads a command from the connection and executes it.
+func (r *RPC) handleRPCConn(conn net.Conn) {
+	defer conn.Close()
+	// RPC connections should execute on the leader.  If we are not the leader,
+	// proxy the connection to the leader so that clients an connect to any node
+	// in the cluster.
+	r.traceCluster("rpc connection from: %v", conn.RemoteAddr())
+
+	if !r.store.IsLeader() {
+		r.proxyLeader(conn.(*net.TCPConn))
+		return
+	}
+
+	// Read and execute request.
+	typ, resp, err := func() (internal.RPCType, proto.Message, error) {
+		// Read request size.
```

Comment:
That's a good idea.  I can do that in a separate PR since this one is getting large.